### PR TITLE
[DAG]SimplifyDemandedVectorElts-add ISD::AVGCEILS/AVGCEILU/AVGFLOORS/AVGFLOORU nodes

### LINKED ISSUE #84768: [DAG] SimplifyDemandedVectorElts - add ISD::AVGCEILS/AVGCEILU/AVGFLOORS/AVGFLOORU nodes
https://rust.godbolt.org/z/e41G5d4q8

As we broadcast the lowest element in `%r0`, the `%s0` shuffle before the avg nodes issuperfluous 
```ll
define <8 x i16> @shadd_demandedelts(<8 x i16> %a0, <8 x i16> %a1) {
  %s0 = shufflevector <8 x i16> %a0, <8 x i16> undef, <8 x i32> zeroinitializer
  %op = call <8 x i16> @llvm.aarch64.neon.shadd.v8i16(<8 x i16> %s0, <8 x i16> %a1)
  %r0 = shufflevector <8 x i16> %op, <8 x i16> undef, <8 x i32> zeroinitializer
  ret <8 x i16> %r0
}
define <8 x i16> @srhadd_demandedelts(<8 x i16> %a0, <8 x i16> %a1) {
  %s0 = shufflevector <8 x i16> %a0, <8 x i16> undef, <8 x i32> zeroinitializer
  %op = call <8 x i16> @llvm.aarch64.neon.srhadd.v8i16(<8 x i16> %s0, <8 x i16> %a1)
  %r0 = shufflevector <8 x i16> %op, <8 x i16> undef, <8 x i32> zeroinitializer
  ret <8 x i16> %r0
}

define <8 x i16> @uhadd_demandedelts(<8 x i16> %a0, <8 x i16> %a1) {
  %s0 = shufflevector <8 x i16> %a0, <8 x i16> undef, <8 x i32> zeroinitializer
  %op = call <8 x i16> @llvm.aarch64.neon.uhadd.v8i16(<8 x i16> %s0, <8 x i16> %a1)
  %r0 = shufflevector <8 x i16> %op, <8 x i16> undef, <8 x i32> zeroinitializer
  ret <8 x i16> %r0
}
define <8 x i16> @urhadd_demandedelts(<8 x i16> %a0, <8 x i16> %a1) {
  %s0 = shufflevector <8 x i16> %a0, <8 x i16> undef, <8 x i32> zeroinitializer
  %op = call <8 x i16> @llvm.aarch64.neon.urhadd.v8i16(<8 x i16> %s0, <8 x i16> %a1)
  %r0 = shufflevector <8 x i16> %op, <8 x i16> undef, <8 x i32> zeroinitializer
  ret <8 x i16> %r0
}
```

### PR DESCRIPTION
**Pull request by aniplcc**
Fixes #84768

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/hadd-combine.ll [hunk 1]
```diff
@@ -879,6 +879,58 @@ define <8 x i16> @uhadd_fixedwidth_v4i32(<8 x i16> %a0, <8 x i16> %a1)  {
   ret <8 x i16> %res
 }
 
+define <8 x i16> @shadd_demandedelts(<8 x i16> %a0, <8 x i16> %a1) {
+; CHECK-LABEL: shadd_demandedelts:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    dup v0.8h, v0.h[0]
+; CHECK-NEXT:    shadd v0.8h, v0.8h, v1.8h
+; CHECK-NEXT:    dup v0.8h, v0.h[0]
+; CHECK-NEXT:    ret
```
Comment (RKSimon) [inline llvm/test/CodeGen/AArch64/hadd-combine.ll:888]: These don't appear to be working? I'd expect:
```
; CHECK:       // %bb.0:
; CHECK-NEXT:    shadd v0.8h, v0.8h, v1.8h
; CHECK-NEXT:    dup v0.8h, v0.h[0]
; CHECK-NEXT:    ret
```

Comment (aniplcc) [reply llvm/test/CodeGen/AArch64/hadd-combine.ll:888]: my bad, forgot to run it against build, will update

Comment (aniplcc) [reply llvm/test/CodeGen/AArch64/hadd-combine.ll:888]: Seems like the addition of the cases don't seem to be doing anything. I don't see any change in the resulting assembly.

**Comment by RKSimon** [comment]
If I had to guess the problem is that we don't hit this until after lowering and the AArch64 DUP node combines aren't calling SimplifyDemandedVectorElts - @davemgreen does that make sense to you?

**Comment by aniplcc** [comment]
Also, the ISD:: opcode passed to SimplifyDemandedVectorElts is `INTRINSIC_WO_CHAIN`(45) rather than one of AVGCEIL(S/U)/AVGFLOOR(S/U)?

**Comment by RKSimon** [comment]
@aniplcc Please can you rebase after 5d3ef06509c2f1fc5384fa64e5848d12f7b8811e and regenerate the test checks?

**Review by RKSimon — APPROVED** [review]
LGTM (keep the AArch64 tests - I'll create a followup issue to improve DUPLANE demanded elts handling)

**Pull request merged**